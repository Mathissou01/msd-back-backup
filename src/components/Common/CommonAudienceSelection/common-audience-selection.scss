@use "src/styles/01-settings/settings.colors" as colors;

.c-CommonAudienceSelection {
  &__NoAudienceAvailableMessage {
    color: colors.$error;
  }

  &__Label {
    color: colors.$blue-2;
    font-weight: 700;
    font-size: 16px;
    line-height: 24px;
  }

  &__List {
    padding-left: 0;
    list-style: none;

    li {
      margin-top: 10px;
    }
  }
}
