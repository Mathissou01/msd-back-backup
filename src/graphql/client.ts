import {
  ApolloClient,
  InMemoryCache,
  NormalizedCacheObject,
} from "@apollo/client";

let client: ApolloClient<NormalizedCacheObject>;

if (process.env.NEXT_PUBLIC_MOCK === "true") {
  /* With .env variable NEXT_PUBLIC_MOCK=true, return mock graphql client with fake response data */
  client = (await import("./mockedClient")).default(1000);
} else {
  client = await new ApolloClient({
    uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql`,
    cache: new InMemoryCache({
      typePolicies: {
        UploadFolder: {
          merge: true,
        },
        UploadFile: {
          merge: true,
        },
        Contract: {
          merge: true,
        },
        ContractCustomization: {
          merge: true,
        },
        New: {
          merge: true,
        },
        Tip: {
          merge: true,
        },
        RecyclingGuideService: {
          merge: true,
        },
        WasteForm: {
          merge: true,
        },
        WasteFamily: {
          merge: true,
        },
        Flow: {
          merge: true,
        },
        FlowColor: {
          merge: true,
        },
        DropOffMap: {
          merge: true,
        },
        Request: {
          merge: true,
        },
        RequestAggregate: {
          merge: true,
        },
      },
    }),
  });
}

export default client;
