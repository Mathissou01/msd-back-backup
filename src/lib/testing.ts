import { prettyDOM } from "@testing-library/react";

export function removeUnstableHtmlPropertiesForTesting(
  htmlElement: HTMLElement,
  attributeName: string,
) {
  const domHTML = prettyDOM(htmlElement, Infinity).toString();
  if (!domHTML) return undefined;
  const regex = new RegExp(`${attributeName}(.*)"(.*)"`, "g");
  return domHTML.replaceAll(regex, "");
}
