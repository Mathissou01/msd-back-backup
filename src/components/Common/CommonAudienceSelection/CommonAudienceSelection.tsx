import React, { useEffect, useRef, useState } from "react";
import { useFormContext } from "react-hook-form";
import { ErrorMessage } from "@hookform/error-message";
import { useGetAudiencesQuery } from "../../../graphql/codegen/generated-types";
import { removeNulls } from "../../../lib/utilities";
import { useContract } from "../../../hooks/useContract";
import { CommonModalWrapperRef } from "../CommonModalWrapper/CommonModalWrapper";
import { IFormSingleMultiselectOption } from "../../Form/FormSingleMultiselect/FormSingleMultiselect";
import CommonButton from "../CommonButton/CommonButton";
import CommonFormErrorText from "../CommonFormErrorText/CommonFormErrorText";
import AudienceModal from "./AudienceModal/AudienceModal";
import "./common-audience-selection.scss";

export default function CommonAudienceSelection() {
  /* Static Data */
  const labels = {
    users: "Usagers *",
    modalOpenningBtn: "SÃ©lectionner les usagers",
    allAudiences: "Tous",
  };
  const requiredMessage = "Ce champ est obligatoire";
  const noAudienceAvailableMessage =
    "Merci d'activer au moins un usager afin de pouvoir enregistrer ce formulaire";

  /* Methods */
  const handleStartModal = () => {
    modalRef.current?.toggleModal(true);
  };

  function setAudiences(audiences: Array<IFormSingleMultiselectOption>) {
    setSelectedAudiences(audiences);
  }

  /* Local Data */
  const modalRef = useRef<CommonModalWrapperRef>(null);
  const { contract } = useContract();
  const {
    watch,
    setValue,
    register,
    formState: { defaultValues, errors },
  } = useFormContext();
  const [selectedAudiences, setSelectedAudiences] =
    useState<Array<IFormSingleMultiselectOption>>();
  const [audienceOptions, setAudienceOptions] =
    useState<Array<IFormSingleMultiselectOption>>();
  const { data: audiencesFromContract, loading: audiencesFromContractLoading } =
    useGetAudiencesQuery({
      variables: {
        filters: {
          contract: {
            id: {
              eq: contract.id,
            },
          },
          isActive: {
            eq: true,
          },
        },
      },
      fetchPolicy: "network-only",
    });
  const audiencesNumberFromContract =
    audiencesFromContract?.audiences?.data?.length;
  const audiencesWatch = watch("audiences");
  register("audiences", {
    required: {
      value: true,
      message: requiredMessage,
    },
  });

  useEffect(() => {
    if (audiencesNumberFromContract === 0) {
      setValue("audiences", undefined);
    } else if (
      audiencesNumberFromContract === 1 &&
      audiencesFromContract?.audiences?.data
    ) {
      const uniqueAssignedAudience = audiencesFromContract.audiences.data[0];
      if (
        uniqueAssignedAudience?.id &&
        uniqueAssignedAudience.attributes?.type
      ) {
        setSelectedAudiences([
          {
            label: uniqueAssignedAudience.attributes.type,
            value: uniqueAssignedAudience.id,
          },
        ]);
      }
    } else if (audiencesNumberFromContract && defaultValues?.audiences) {
      const filteredAudiences = defaultValues.audiences.filter(
        (audience: IFormSingleMultiselectOption) =>
          audiencesFromContract?.audiences?.data?.some(
            (audience2) => audience2.id === audience.value,
          ),
      );
      setSelectedAudiences(filteredAudiences);
    }
  }, [
    audiencesNumberFromContract,
    defaultValues,
    setValue,
    audiencesFromContract?.audiences?.data,
  ]);

  useEffect(() => {
    if (audiencesWatch === undefined) {
      register("audiences", {
        required: {
          value: true,
          message: requiredMessage,
        },
      });
    }
  }, [audiencesWatch, register]);

  useEffect(() => {
    if (audiencesFromContract?.audiences?.data && !audienceOptions) {
      setAudienceOptions(
        audiencesFromContract.audiences.data
          .map((audience) => {
            if (audience?.attributes?.type && audience.id) {
              return {
                label: audience.attributes.type,
                value: audience.id,
              };
            }
          })
          .filter(removeNulls),
      );
    }
  }, [audiencesFromContract?.audiences?.data, audienceOptions]);

  return (
    <>
      <div className="c-CommonAudienceSelection__Label">{labels.users}</div>
      {!audiencesFromContractLoading && (
        <>
          {audiencesNumberFromContract === 0 ? (
            <span className="c-CommonAudienceSelection__NoAudienceAvailableMessage">
              {noAudienceAvailableMessage}
            </span>
          ) : (
            <ul className="c-CommonAudienceSelection__List">
              {selectedAudiences &&
                (selectedAudiences.length > 1 ? (
                  selectedAudiences?.length === audiencesNumberFromContract ? (
                    <li>{labels.allAudiences}</li>
                  ) : (
                    selectedAudiences.map((audience, index) => {
                      return <li key={index}>{audience.label}</li>;
                    })
                  )
                ) : (
                  selectedAudiences.length === 1 && (
                    <li>{selectedAudiences[0].label}</li>
                  )
                ))}
            </ul>
          )}
          <CommonButton
            label={labels.modalOpenningBtn}
            type="button"
            style="secondary"
            isDisabled={
              !audiencesNumberFromContract || audiencesNumberFromContract <= 1
            }
            onClick={() => handleStartModal()}
          />
          <ErrorMessage
            errors={errors}
            name="audiences"
            render={({ message }: { message: string }) => (
              <CommonFormErrorText
                message={message}
                errorId={`audiences_error`}
              />
            )}
          />
        </>
      )}
      <AudienceModal
        modalRef={modalRef}
        onValidate={setAudiences}
        selectedAudiences={selectedAudiences}
        audienceOptions={audienceOptions ?? []}
      />
    </>
  );
}
