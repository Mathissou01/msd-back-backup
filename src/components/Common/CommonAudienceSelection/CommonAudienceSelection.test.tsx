import { fireEvent, render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import { FormProvider, useForm } from "react-hook-form";
import { ReactNode } from "react";
import {
  getAllAudiencesMockedData,
  getNoAudienceMockedData,
  getSingleAudienceMockedData,
  getTwoAudiencesMockedData,
} from "../../../../__mocks__/audiencesMockData";
import CommonAudienceSelection from "./CommonAudienceSelection";

jest.mock("../../../graphql/client", () => null);

const audienceModalTitle = "Sélectionner les usagers";
const openModalButtonLabel = "Sélectionner les usagers";
const modalSubmitBtnLabel = "Valider la sélection";

const individualsText = "Particuliers";
const collectivesText = "Collectifs";
const twoSelectedAudiencesValue = {
  audiences: [
    { label: "Particuliers", value: 1 },
    { label: "Collectifs", value: 2 },
  ],
};
const allAudiencesText = "Tous";
const noAudienceAvailableMessage =
  "Merci d'activer au moins un usager afin de pouvoir enregistrer ce formulaire";

const Wrapper = (props: { children: ReactNode }) => {
  const formMethods = useForm();
  return (
    <>
      <FormProvider {...formMethods}>{props.children}</FormProvider>{" "}
      <div id="modal-portal" />
    </>
  );
};

const WrapperWithTwoSelectedAudiences = (props: { children: ReactNode }) => {
  const formMethods = useForm({
    defaultValues: twoSelectedAudiencesValue,
  });
  return (
    <>
      <FormProvider {...formMethods}>{props.children}</FormProvider>
      <div id="modal-portal" />
    </>
  );
};

describe("CommonAudienceSelection", () => {
  it("renders", () => {
    const { container } = render(
      <Wrapper>
        <MockedProvider mocks={getSingleAudienceMockedData}>
          <CommonAudienceSelection />
        </MockedProvider>
      </Wrapper>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders without audiences", async () => {
    const { container } = render(
      <Wrapper>
        <MockedProvider mocks={getNoAudienceMockedData}>
          <CommonAudienceSelection />
        </MockedProvider>
      </Wrapper>,
    );
    const noAudienceAvailableText = await screen.findByText(
      noAudienceAvailableMessage,
    );
    expect(noAudienceAvailableText).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("has working behaviour when having single audience and clicking button", async () => {
    const { container } = render(
      <Wrapper>
        <MockedProvider mocks={getSingleAudienceMockedData}>
          <CommonAudienceSelection />
        </MockedProvider>
      </Wrapper>,
    );
    const audienceModalContent = screen.queryByText(audienceModalTitle);
    expect(audienceModalContent).toBeNull();
    const button = await screen.findByText(openModalButtonLabel);
    expect(button).toBeInTheDocument();
    expect(await screen.findByText(individualsText)).toBeInTheDocument();
    await fireEvent.click(button);
    expect(await screen.findByText(audienceModalTitle)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("has working behaviour when having two audiences and clicking button", async () => {
    render(
      <Wrapper>
        <MockedProvider mocks={getTwoAudiencesMockedData}>
          <CommonAudienceSelection />
        </MockedProvider>
      </Wrapper>,
    );
    const audienceModalContent = screen.queryByText(audienceModalTitle);
    expect(audienceModalContent).toBeNull();
    const selectUsersButton = await screen.findByText(openModalButtonLabel);
    expect(selectUsersButton).toBeInTheDocument();
    expect(screen.queryByText(individualsText)).not.toBeInTheDocument();
    await fireEvent.click(selectUsersButton);
    expect(await screen.findAllByText(audienceModalTitle)).toHaveLength(2);
  });

  it("has working behaviour when having two audiences already selected matching all available audiences", async () => {
    const { container } = render(
      <WrapperWithTwoSelectedAudiences>
        <MockedProvider mocks={getTwoAudiencesMockedData}>
          <CommonAudienceSelection />
        </MockedProvider>
      </WrapperWithTwoSelectedAudiences>,
    );
    const selectUsersButton = await screen.findByText(openModalButtonLabel);
    expect(selectUsersButton).toBeInTheDocument();
    expect(await screen.findByText(allAudiencesText)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("has working behaviour when having two audiences already selected not matching all available audiences", async () => {
    const { container } = render(
      <WrapperWithTwoSelectedAudiences>
        <MockedProvider mocks={getAllAudiencesMockedData}>
          <CommonAudienceSelection />
        </MockedProvider>
      </WrapperWithTwoSelectedAudiences>,
    );
    const selectUsersButton = await screen.findByText(openModalButtonLabel);
    expect(selectUsersButton).toBeInTheDocument();
    expect(await screen.findByText(individualsText)).toBeInTheDocument();
    expect(await screen.findByText(collectivesText)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("has working behaviour when having two audiences selected and clicking on submitButton in AudienceModal", async () => {
    const { container } = render(
      <WrapperWithTwoSelectedAudiences>
        <MockedProvider mocks={getTwoAudiencesMockedData}>
          <CommonAudienceSelection />
        </MockedProvider>
      </WrapperWithTwoSelectedAudiences>,
    );
    const selectUsersButton = await screen.findByText(openModalButtonLabel);
    expect(selectUsersButton).toBeInTheDocument();
    await fireEvent.click(selectUsersButton);
    const modalSubmitButton = await screen.findByText(modalSubmitBtnLabel);
    expect(modalSubmitButton).toBeInTheDocument();
    await fireEvent.click(modalSubmitButton);
    expect(await screen.findByText(allAudiencesText)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
