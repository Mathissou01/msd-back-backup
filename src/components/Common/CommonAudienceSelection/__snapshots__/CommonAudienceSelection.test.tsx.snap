// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommonAudienceSelection has working behaviour when having single audience and clicking button 1`] = `
<div>
  <div
    class="c-CommonAudienceSelection__Label"
  >
    Usagers *
  </div>
  <ul
    class="c-CommonAudienceSelection__List"
  >
    <li>
      Particuliers
    </li>
  </ul>
  <button
    class="c-CommonButton c-CommonButton_disabled c-CommonButton_secondary c-CommonButton_fontSmall"
    disabled=""
    type="button"
  >
    Sélectionner les usagers
  </button>
   
  <div
    id="modal-portal"
  />
</div>
`;

exports[`CommonAudienceSelection has working behaviour when having two audiences already selected matching all available audiences 1`] = `
<div>
  <div
    class="c-CommonAudienceSelection__Label"
  >
    Usagers *
  </div>
  <ul
    class="c-CommonAudienceSelection__List"
  >
    <li>
      Tous
    </li>
  </ul>
  <button
    class="c-CommonButton c-CommonButton_secondary c-CommonButton_fontSmall"
    type="button"
  >
    Sélectionner les usagers
  </button>
  <div
    id="modal-portal"
  />
</div>
`;

exports[`CommonAudienceSelection has working behaviour when having two audiences already selected not matching all available audiences 1`] = `
<div>
  <div
    class="c-CommonAudienceSelection__Label"
  >
    Usagers *
  </div>
  <ul
    class="c-CommonAudienceSelection__List"
  >
    <li>
      Particuliers
    </li>
    <li>
      Collectifs
    </li>
  </ul>
  <button
    class="c-CommonButton c-CommonButton_secondary c-CommonButton_fontSmall"
    type="button"
  >
    Sélectionner les usagers
  </button>
  <div
    id="modal-portal"
  />
</div>
`;

exports[`CommonAudienceSelection has working behaviour when having two audiences selected and clicking on submitButton in AudienceModal 1`] = `
<div>
  <div
    class="c-CommonAudienceSelection__Label"
  >
    Usagers *
  </div>
  <ul
    class="c-CommonAudienceSelection__List"
  >
    <li>
      Tous
    </li>
  </ul>
  <button
    class="c-CommonButton c-CommonButton_secondary c-CommonButton_fontSmall"
    type="button"
  >
    Sélectionner les usagers
  </button>
  <div
    id="modal-portal"
  />
</div>
`;

exports[`CommonAudienceSelection renders 1`] = `
<div>
  <div
    class="c-CommonAudienceSelection__Label"
  >
    Usagers *
  </div>
   
  <div
    id="modal-portal"
  />
</div>
`;

exports[`CommonAudienceSelection renders without audiences 1`] = `
<div>
  <div
    class="c-CommonAudienceSelection__Label"
  >
    Usagers *
  </div>
  <span
    class="c-CommonAudienceSelection__NoAudienceAvailableMessage"
  >
    Merci d'activer au moins un usager afin de pouvoir enregistrer ce formulaire
  </span>
  <button
    class="c-CommonButton c-CommonButton_disabled c-CommonButton_secondary c-CommonButton_fontSmall"
    disabled=""
    type="button"
  >
    Sélectionner les usagers
  </button>
   
  <div
    id="modal-portal"
  />
</div>
`;
